sys/net/gnrc/routing/simple_subnets/simple_subnets.c

```diff
@@ -0,0 +1,115 @@
+ * the process repeats until the bits of prefix are exhausted. (The smallest
+ * must still have a /64 prefix.)
+ *
+ * The downstream router will send an otherwise empty router advertisement with only
+ * the Route Information Option set to the upstream network.
+ * The Route Information Option contains the prefix of the downstream network so that
+ * upstream hosts will no longer consider hosts in this subnet on-link but instead
+ * will use the downstream router to route to the new subnet.
+ *
+ * Usage
+ * =====
+ *
+ * Simply add the `gnrc_simple_subnets` module to the nodes that should act as routers
+ * in the cascading network.
+ * The upstream network will be automatically chosen as the one that first receives a
+ * router advertisement.
+ *
+ * @{
+ *
+ * @file
+ * @author  Benjamin Valentin <benjamin.valentin@ml-pa.com>
+ */
+
+#include "net/gnrc/ipv6/nib.h"
+
+#define ENABLE_DEBUG 0
+#include "debug.h"
+
+static char addr_str[IPV6_ADDR_MAX_STR_LEN];
+
+void gnrc_nib_rtr_adv_pio_cb(gnrc_netif_t *upstream, const ndp_opt_pi_t *pio)
+{
+    gnrc_netif_t *downstream = NULL;
+    const ipv6_addr_t *prefix = &pio->prefix;
+    uint32_t valid_ltime = byteorder_ntohl(pio->valid_ltime);
+    uint32_t pref_ltime = byteorder_ntohl(pio->pref_ltime);
+
+    /* create a subnet for each downstream interface */
+    unsigned subnets = gnrc_netif_numof() - 1;
+
+    const uint8_t prefix_len = pio->prefix_len;
+    uint8_t new_prefix_len;
+
+    if (subnets == 0) {
+        return;
+    }
+
+    new_prefix_len = prefix_len + 32 - __builtin_clz(subnets);
+
+    if (new_prefix_len > 64) {
+        DEBUG("nib: can't split /%u into %u subnets\n", prefix_len, subnets);
+        return;
+    }
+
+    while ((downstream = gnrc_netif_iter(downstream))) {
+        ipv6_addr_t new_prefix;
+
+        if (downstream == upstream) {
+            continue;
+        }
+
+        /* create subnet by adding interface index */
+        new_prefix.u64[0].u64 = byteorder_ntohll(prefix->u64[0]);
+        new_prefix.u64[0].u64 |= (uint64_t)subnets-- << (63 - prefix_len);
+        new_prefix.u64[0] = byteorder_htonll(new_prefix.u64[0].u64);
+
+        DEBUG("nib: configure prefix %s/%u on %u\n",
+              ipv6_addr_to_str(addr_str, &new_prefix, sizeof(addr_str)),
+              new_prefix_len, downstream->pid);
+
+        gnrc_netif_ipv6_add_prefix(downstream, &new_prefix, new_prefix_len,
+                                   valid_ltime, pref_ltime);
+
+        /* start advertising subnet */
+        gnrc_ipv6_nib_change_rtr_adv_iface(downstream, true);
+    }
+
+    /* Disable router advertisements on upstream interface.
+     * 1. Does not confuse the upstream router to add the border router to its
+     *    default router list and
+     * 2. Includes Route Information Option in last router advertisement to
+     *    inform upstream router about downstream subnets.
+     */
+    gnrc_ipv6_nib_send_final_rtr_adv(upstream);
```

Comment:
```suggestion
    /* Immediately send an RA with RIO
     */
    gnrc_pktsnip_t *ext_opts = NULL;

    while (/* use gnrc_ipv6_nib_pl_iter to get PL entries */) {
            gnrc_pktsnip_t *snip  = gnrc_ndp_opt_ri_build(&ple.pfx,
                                                          ple.pfx_len,
                                                          valid_ltime,
                                                          NDP_OPT_RI_FLAGS_PRF_NONE,
                                                          ext_opts);
            if (snip != NULL) {
                ext_opts = snip;
            } else {
                break;
            }
    }

    gnrc_ndp_rtr_adv_send(netif, NULL, &ipv6_addr_all_nodes_link_local,
                          true, ext_opts);
```

Reply:
even saves us iterating over all PL entries twice

Reply:
See https://github.com/RIOT-OS/RIOT/pull/16706 for how to get the `valid_ltime` ;-).

Reply:
> See #16706 for how to get the `valid_ltime` ;-).

Ah, just saw that you do not even need it. Well, I think #16706 still has value for the reasons pointed out there.

Reply:
Yes lifetime output was something I was sorely missing from the `nib` commands, now we can have that too!